Add mussel

firasuke opened the pull request:
mussel is the shortest and fastest script available today to build working cross compilers that target musl libc.

## Features
1. **Up-to-date**: uses latest available upstream sources for packages
2. **Fast**: probably the fastest script around to build a cross compiler
   targetting musl libc, also it's written entirely in POSIX DASH
3. **Short**: has the least amount of steps (see below) required to build a
   cross compiler targetting musl libc (even less than
   [musl-cross-make](https://github.com/richfelker/musl-cross-make))
4. **Small**: all installation steps use `install-strip` when applicable
5. **Simple**: easy to read, modify and extend
6. **POSIX Compliant**: written entirely in POSIX DASH
7. **Well Documented**: the script has comments (that are considered state of
   the art information) all over the place explaining what is being done and why

richfelker commented:
I'm skeptical of the claim that it's fastest (doesn't look like there's any reason it should be faster than mcm) and it's doing some things that definitely produce a broken libc like `-ffast-math` if they're not caught (not sure how they're uncaught; configure is supposed to error out if you do that).

firasuke commented:
The check for `-ffast-math` is being patched from the configure file to allow musl to be built without depending on libgcc for certain archs like `x86-64`.

richfelker commented:
I'm not opposed to adding this but it should be described as a WIP with possible bugs, not in a way that implies it's the recommended way to get a cross toolchain.

richfelker commented:
The patch allowing `-ffast-math` is dangerously wrong and produces a broken libc. The idea that you can build without a libgcc is wrong. This is not possible for most archs. And the procedure you're using to get a libgcc on the ones that do need it is far slower and costlier than it needs to be. It looks like you're also bootstrapping the first configure of musl with the host gcc, so this isn't even actually a real cross-compiler; it necessarily has to be for the same arch+ABI as the existing host.

protonesso commented:
kek

firasuke commented:
The general approach to build a cross compiler that uses gcc and targets musl libc is to:
1- configure, build and install `cross binutils`
2- configure, build  and install a small static cross gcc + libgcc (libgcc.a, not linked against any C library)
3- configure, build and install musl
4- configure, build and install our final cross gcc
5- build and install libgcc (linked against musl)
6- build and install libstdc++-v3 (for C++ support)
7- build and install libgomp (for OpenMP support)

The problem with this approach is the need to build GCC twice, which takes a lot of time. The first pass is basically just a stub to be able to configure musl.

So musl needs libgcc, and the only way to provide it without taking much time is by libgcc.a, and libgcc also needs musl to link to it so there's a circular dependency.

In mussel, we're providing musl headers beforehand to be able to configure, build and install our final cross gcc (without building libgcc since musl isn't available yet).

To be able to configure musl to only install its headers, we must use the host's C compiler at first, and remove the dependency on the non existent libgcc, since all we want is just the headers. So we pass `CC=gcc` and remove the check for `-ffast-math` to be able to do that.

Once we have our working cross gcc compiler, we build and install musl using `$XTARGET-gcc` (simply by changing the `CC` variable inside musl's config.mak that resulted from configuring `musl-headers`).

Then we build and install our cross compiler's `libgcc` linked against `musl`, as well as `libstdc++-v3` and `libgomp` if needed.

So libgcc will still be available to all archs, and will still be linked against musl for all archs. Some archs like `x86-64`, `powerpc64` and `powerpc64le` can do without the `libgcc.a` when building musl, others (and they are a lot of archs) require `libgcc.a` so we still provide it as well.

So we'll still providing `libgcc` for all archs, it's just that for certain archs it can be built once (And be linked against musl), and for others it is built twice to satisfy musl.

firasuke commented:
> And the procedure you're using to get a libgcc on the ones that do need it is far slower and costlier than it needs to be.
How is it far slower and costlier, if it's done in less steps and requires less time and produces the exact same results?

richfelker commented:
> The problem with this approach is the need to build GCC twice, which takes a lot of time. The first pass is basically just a stub to be able to configure musl.

musl-cross-make already solved this problem. It only builds gcc once.

> To be able to configure musl to only install its headers, we must use the host's C compiler at first,

This is possible only if host and target arch+ABI match, which is a very specific case. On powerpc[64]-glibc it's *never* the case, as you found out. However you could make it work by passing `-mlong-double-64` and the secureplt options as part of `CC` to override the mismatched ABI on the host. Instead you're overriding the checks for wrong ABI and producing a broken musl...

> and remove the dependency on the non existent libgcc, since all we want is just the headers.

`configure` should not depend on libgcc anyway. Neither does compiling a static `libc.a`. The only place libgcc is needed in the musl build process is for *linking* `libc.so`.

> So we pass CC=gcc and remove the check for -ffast-math to be able to do that.

This also produces a broken libc.

firasuke commented:
> musl-cross-make already solved this problem. It only builds gcc once.

`musl-cross-make` is overly complicated for what it tries to solve.

> This is possible only if host and target arch+ABI match, which is a very specific case. On powerpc[64]-glibc it's _never_ the case, as you found out. However you could make it work by passing `-mlong-double-64` and the secureplt options as part of `CC` to override the mismatched ABI on the host. Instead you're overriding the checks for wrong ABI and producing a broken musl...

A step to reconfigure `musl` could be added if that's the case.

> `configure` should not depend on libgcc anyway. Neither does compiling a static `libc.a`. The only place libgcc is needed in the musl build process is for _linking_ `libc.so`.

Again, adding a reconfigure step to `musl` after cross-gcc have been built would also work.

> This also produces a broken libc.

Again, a second reconfiguration step would prevent the need to manually edit `CC` and `LIBCC` in config.mak, and would probably remove the need for `-ffast-math`.

firasuke commented:
I don't mind adding `WIP` to mussel in its current state, and adding an additional step to reconfigure `musl` to fix some ABI breakages, but for what it's being used currently, I've gotten identical results from both `mcm` and `mussel` having built over `80` packages with them for the listed supported archs.

richfelker commented:
> `musl-cross-make` is overly complicated for what it tries to solve.

I can only imagine this is a matter of your not liking declarative form. If you want to flatten the same idea as a procedural script, that works too, and it will be at least as simple as what you're doing.

> A step to reconfigure musl could be added if that's the case.

This would probably fix the serious problems. But I don't see why you think you need `-ffast-math` at all since it makes no difference if all you're running is `install-headers`. If you're actually compiling source files, it will produce broken object files.

> I've gotten identical results from both mcm and mussel having built over 80 packages with them for the listed supported archs.

I'm pretty sure that if you put some effort into testing math functions, `strtod`, etc. you'd find that they're producing incorrect values. Probably even for cases where there's an exact result. You've also got broken powerpc binaries with mismatched ABI.

In addition to these major issues, there are a number of smaller fixes in the mcm gcc patches that you're missing, like invalid TLS model being used for libgomp, etc. There's also the riscv TLS copy relocs issue (whereby gcc may produce binaries that won't work with musl's dynamic linker), but it might be fixed upstream now in gcc 10 since you've got that (mcm is still missing it).

firasuke commented:
> I can only imagine this is a matter of your not liking declarative form. If you want to flatten the same idea as a procedural script, that works too, and it will be at least as simple as what you're doing.

Yes, I originally planned to do that as you can see [here](https://gist.github.com/firasuke/82ee9ed032aaf61af26826e07a3db14d).

> This would probably fix the serious problems. But I don't see why you think you need `-ffast-math` at all since it makes no difference if all you're running is `install-headers`. If you're actually compiling source files, it will produce broken object files.

I used to have this step originally for mussel but removed it thinking that a single musl configuration in `musl-headers` would produce similar results. Based on my testing, it produced similar results.

> I'm pretty sure that if you put some effort into testing math functions, `strtod`, etc. you'd find that they're producing incorrect values. Probably even for cases where there's an exact result. You've also got broken powerpc binaries with mismatched ABI.

I see.

> In addition to these major issues, there are a number of smaller fixes in the mcm gcc patches that you're missing, like invalid TLS model being used for libgomp, etc. There's also the riscv TLS copy relocs issue (whereby gcc may produce binaries that won't work with musl's dynamic linker), but it might be fixed upstream now in gcc 10 since you've got that (mcm is still missing it).

I was planning on adding these later on once I got the thing to work. Since `libgomp` is currently optional, it wasn't tested as much, so I'll be sure to use relevant patches from `musl-cross-make`.

Also, why does `musl-cross-make` use these [old headers](https://github.com/sabotage-linux/kernel-headers)?

richfelker commented:
> Also, why does musl-cross-make use these old [kernel] headers?

At the last time the kernel headers version was updated, there was no upstream kernel version that had fully time64-compatible headers. I believe the last header fixes needed appeared in 5.6. Now we should be able to move to unpatched 5.6 or later.

Note that the main distros testing at the time were not using bleeding-edge kernels but LTS, and were using the same kernel source version for headers and actual kernel, so it made sense to be building on that LTS. And in fact it was no harder to make the patches against it than against 5.4 (which was the just-released LTS at the time). Either way the same changed were needed to fix the time64 incompatibilities.

firasuke commented:
I see, thanks for taking the time to clearly explain your thoughts.

firasuke commented:
Ok I've updated `mussel` to fix the ABI breakage:
- Removed `-ffast-math`, and the two powerpc "ABI-breaking" patches
- Building libgcc-static is now done for all archs before musl
- Reconfigure musl again just before building it

It's also still a single pass to build GCC.

firasuke commented:
Please proceed to merge this as all the proposed changes have been committed.

firasuke commented:
Any updates on this as to why it hasn't been merged yet?

Thanks.

somasis commented:
Sorry for the delay, lost track of this PR. I'd just like a go ahead from Rich before it gets merged, as I don't really feel confident giving my input on this. @richfelker

richfelker commented:
I've merged it (339f00969b38f48d56ff35a3b6af57cd807862a3) without the unverified promotional claim in the commit message.

firasuke commented:
Thanks for your time and effort.
